crates/bevy_math/src/cubic_splines.rs

```diff
@@ -239,14 +264,39 @@ impl<P: VectorSpace> CubicGenerator<P> for CubicCardinalSpline<P> {
         let mirrored_last = self.control_points[length - 1] * 2. - self.control_points[length - 2];
         let extended_control_points = once(&mirrored_first)
             .chain(self.control_points.iter())
-            .chain(once(&mirrored_last))
-            .collect::<Vec<_>>();
+            .chain(once(&mirrored_last));
 
         let segments = extended_control_points
-            .windows(4)
-            .map(|p| CubicSegment::coefficients([*p[0], *p[1], *p[2], *p[3]], char_matrix))
+            .tuple_windows()
+            .map(|(&p0, &p1, &p2, &p3)| {
+                CubicSegment::coefficients([p0, p1, p2, p3], self.char_matrix())
+            })
+            .collect();
+
+        CubicCurve { segments }
+    }
+}
+impl<P: VectorSpace> CyclicCubicGenerator<P> for CubicCardinalSpline<P> {
+    #[inline]
+    fn to_curve_cyclic(&self) -> CubicCurve<P> {
+        // Unlike in `to_curve`, here we don't augment the control points at all, instead just reusing the
+        // "wrapped" control points. Note that this guarantees that the differentiability guarantees are
+        // satisfied automatically; the curve passes through every pair of adjacent control points, including
+        // the start and end, and wraps with an end -> start segment.
+        let mut segments: Vec<CubicSegment<P>> = self
+            .control_points
+            .iter()
+            .circular_tuple_windows()
+            .map(|(&p0, &p1, &p2, &p3)| {
+                CubicSegment::coefficients([p0, p1, p2, p3], self.char_matrix())
+            })
             .collect();
 
+        // Rotate the list of segments so that the curve has the expected parametrization — i.e., we want
+        // the first curve segment to be the one connecting the first two control points, but to begin with
+        // it was the segment connecting the second and third.
+        segments.rotate_left(1);
```

Comment:
Is this a copy-ing in-place rotation?
Can this be avoided by somehow placing this before the `.collect()`?

Probably not worth it if it makes the code more nasty.

Reply:
Yeah, this is unfortunately an O(n) operation in the number of segments. I was lazy in writing it this way, and probably we can do better by chaining some iterators together and only taking the correct number of segments. That would read more confusingly but be more efficient (probably.

Reply:
I rewrote it to avoid doing anything like this; now it picks out the last segment individually and uses the iterator to make the rest of the curve. It's uglier, but it looks like it might be faster. Let me know what you think; we can always revert it if we need to.